packages/contracts-rfq/contracts/FastBridgeV2.sol

```diff
@@ -131,28 +159,33 @@ contract FastBridgeV2 is Admin, IFastBridgeV2, IFastBridgeV2Errors {
     }
 
     /// @inheritdoc IFastBridgeV2
+    // TODO: reduce cyclomatic complexity alongside arbitrary call
+    // solhint-disable-next-line code-complexity
     function relay(bytes memory request, address relayer) public payable {
         if (relayer == address(0)) revert ZeroAddress();
+        // Check if the transaction has already been relayed
         bytes32 transactionId = keccak256(request);
-        BridgeTransaction memory transaction = getBridgeTransaction(request);
-        if (transaction.destChainId != uint32(block.chainid)) revert ChainIncorrect();
-
-        // check haven't exceeded deadline for relay to happen
-        if (block.timestamp > transaction.deadline) revert DeadlineExceeded();
-
-        if (bridgeRelayDetails[transactionId].relayer != address(0)) revert TransactionRelayed();
-
+        if (bridgeRelays(transactionId)) revert TransactionRelayed();
+        // Decode the transaction and check that it could be relayed on this chain
+        BridgeTransactionV2 memory transaction = getBridgeTransactionV2(request);
+        if (transaction.txV1.destChainId != uint32(block.chainid)) revert ChainIncorrect();
+        // Check the deadline for relay to happen
+        if (block.timestamp > transaction.txV1.deadline) revert DeadlineExceeded();
+        // Check the exclusivity period, if it is still ongoing
+        if (block.timestamp <= transaction.exclusivityEndTime && relayer != transaction.exclusivityRelayer) {
```

Comment:
it seems like it might be possible for this to briefly delay a non-exclusive relay from occurring due to clock gaps btwn chains.

That is, if the dest chain is X seconds behind the origin, it would not be possible for anyone to relay until the EndTime is reached on the dest, no?

Should we restructure the checks like this:

if exclusivityRelayer is 0 or `relayer`, do not bother comparing exclusivity timestamp & just move along.

Otherwise, proceed to check that the exclusivity end time is in the past & revert if not.


This might also be more gas efficient overall

Reply:
For "non-exclusive txs" `exclusivityEndTime == 0` (set by using `paramsV2.quoteExclusivitySeconds = 0`), so this will actually succeed without checking the relayer address.

However, the dest chain being behind would mean that the exclusivity period is extended longer than it should be for "exclusive txs". I guess we could treat this in two ways:
1. Leave as is, and use much lower values on the FE for such cases (e.g. 1 second vs default value of 30 seconds).
2. Allow negative values for `quoteExclusivitySeconds` so that the `exclusivityEndTime` could be set to a past timestamp origin-chain-wise, but a future timestamp for a destination chain, which is behind. In this casethe FE would use a negative value for such scenarios.

These are mostly similar, with the only difference being setting "exclusivity seconds" to either one second, or to a negative value (setting it to zero will turn off the exclusivity).

What do you think about these options?

Reply:
Ah I was neglecting this piece of bridge() and instead thinking that a non-excl would just have an endtime +0 seconds from deposit block ts
`        uint256 exclusivityEndTime =
            paramsV2.quoteExclusivitySeconds == 0 ? 0 : block.timestamp + paramsV2.quoteExclusivitySeconds;`
            


option 2 seems like the better choice -- offers more flexibility & seems like it should not cost anything extra to allow a negative offset. If wanted to also allow zero exclusive seconds, could we switch to exclRelayer address = 0 to indicate/detect that exclusivity is off?

Reply:
Could go for `quoteExclusivitySeconds` being `int256` rather than `uint256` to allow negative values, don't see any concerns with that (txs having `exclusivityEndTime` in the past could be still completed by anyone).

As for allowing zero exclusive seconds with a quote relayer address, I think this is a very niche usecase, which might as well use +-1 instead of zero. Having this check in place allows to prevent bridge txs that mistakenly have only one exclusivity param from being initiated on the origin chain - this looks like a bigger concern to me than having to use +-1 instead of zero.

Reply:
Actually, I take back the comment above. Having `quoteRelayer` as a single indicator of whether exclusivity is requested feels simple enough, and the incorrect integration (setting quote relayer but not seconds) isn't actually that bad - it's easily solvable and doesn't put anything at risk.

And this is also easier to enforce in the contracts. Will rework the tests and follow up with the contract changes